lightning/src/routing/router.rs

```diff
@@ -1498,6 +1501,55 @@ where L::Target: Logger {
 		}
 	}
 
+	// Step (10).
+	// Add a random 'shadow route' offset to the CLTV expiry delta values to improve privacy.
+	//
+	// Start from the payee
+	let mut cur_node_id = payee_node_id;
+	let mut shadow_ctlv_expiry_delta_offset: u32 = 0;
+
+	let mut prng = ChaCha20::new(random_bytes, &[0; 8]);
+	let mut random_index_bytes = [0u8; 8];
+	let mut random_index: usize;
+
+	// Do a three-hop random walk
+	for _random_hop in 0..3 { 
```

Comment:
Should we randomize the number here instead of always using 3?

Reply:
We definitely could draw it from [0..3] or so, but I'm not entirely sure what would be gained by this. I chose length 3 simply because it roughly reflects current avg. path lengths in the network and therefore should offer an anonymity set corresponding to most of the network.

Reply:
I worry just always picking 3 reduces our anonymity set - instead of a node knowing the recipient is the last hop, they now know that the recipient is 3 before the last, as calculated using CLTV. Sure, we pick a plausible last-hop, but there may only be one option with the exact CLTV delta we pick. If we pick in the range [1..3] then even if there's only one plausible path with the exact CLTV we use an intermediary still only learns that the recipient is one of 3 nodes.

Reply:
Hm, I think since backtracking and loops aren't prevented during the random walk and basically the whole network should be reachable in 3 hops, the anonymity set should not be all that different to begin with. However, I agree that the randomization may least not hurt and is likely useful also for another reason: it may allow us increase/maintain uncertainty while keeping the added 'shadow route' shorter and the CTLV offset lower. This is likely something we want since adding a larger offset may increase the payment failure probability.

Reply:
> basically the whole network should be reachable in 3 hops

Ah, that's a good point. I guess my concern was that we're picking specific CLTV values based on real nodes, and if we happen to pick three nodes that have something other than super common CLTVs then we're basically announcing who we are again. We could do something much fancier to try to reduce this likelihood, or we could just do 1-3 instead of 3 :)

Reply:
Implemented the randomization with https://github.com/lightningdevkit/rust-lightning/commit/ea0ad9d079ba51e5cf6f780999661a1b7673351b. The question on a reasonable upper-limit for the offset remains for me.

Reply:
Yea, I dunno what the answer is, common node CLTV difference values are 144, right? On the upper end? Given we'd prefer to just use "network-common values" we should just limit to 3*144, no? We could also apply no limit and rely on the `PaymentParameters::max_total_cltv_expiry_delta` setting, which defaults to the total path being limited to a full week.

Reply:
I just had another look at ClTV value data I had attached to #1234: the most prevalent value by far is 40 (count: 116133), with 144 only being second (count: 9608). However, as 144 is also our default, I'll probably go with 3*144 for the limit, even though it might technically provide a smaller anonymity set. I think relying on `max_total_cltv_expiry_delta` is not an option here, since this limit is checked and enforced during path finding, i.e., before we add the offset. When we don't have a limit, we might just risk that the payment might get rejected by intermediates during routing itself.

Reply:
> I think relying on max_total_cltv_expiry_delta is not an option here, since this limit is checked and enforced during path finding, i.e., before we add the offset

Right, but we do need to ensure we never violate `max_total_cltv_expiry_delta` on the full route, including the offset, as that's the guarantee we made to the user. Maybe we should reduce the max we use during pathfinding by 2*144 and then refuse to overrun the cap when we add the offset?

Reply:
Well if the median is 40, we can also just reduce the cap by 2*40 and then go from there? That way we're at least being relatively conservative.

Reply:
Alright, but then we probably also want to use 40 also as the default fallback value for private paths, which would increase the likelihood to stay below the offset limit and the total limit.

Reply:
https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/3794bf60245095e06dce60f2babe4e46e6920ecd implements what we discussed so far: 1) subtract `2*40` from the `max_total_cltv_expiry_delta` limit to account for the offset added later. 2) limit the max offset to `3*144`. 3) When we're handling a entirely private path, we add a random offset `x*40, x in [1,...3]`.

Two comments: 1) wouldn't it be 'more conservative' to exclude more failure possibilities from the beginning, i.e., reduce the `max_total_cltv_expiry_delta` value used during path finding by`2*144`? 2) In the case that the CLTV delta  exceeds the total limit after we add the offset, the pathfinding currently just fails with a `LightningError`. Maybe we should at least make this happen less often by applying the random offset and the check not on the final `selected_paths`, but rather before on `drawn_routes`, i.e., swapping Step (9) and Step (10)?

Reply:
> 1) wouldn't it be 'more conservative' to exclude more failure possibilities from the beginning, i.e., reduce the max_total_cltv_expiry_delta value used during path finding by2*144? 2) In the case that the CLTV delta exceeds the total limit after we add the offset, the pathfinding currently just fails with a LightningError

Ah, I'd been thinking that instead of failing at that point we'd simply cap how much additional CLTV we add. ie if we find a path that is exactly 2*40 less than our `max_total_cltv_expiry_delta` we'd cap our shadow route offset at 80, even if we pick a shadow route higher.

Reply:
Ah I see, so the random offset would only be applied when there is still 'room left' beneath the `max_total_cltv_expiry_delta` limit. The issue I see with that (and with enforcing a fixed limit in general) is that there may be some paths for which we only ever can add one or none hop's worth of randomness, which might significantly reduce the anonymity set. In particular in the case where no offset is added at all, the second to last hop on the path could infer that it's successor is the payee, either because the remainder exactly matches the publicly announced `cltv_expiry_delta`, or because it is so small that it's improbable that another hop 'fits in there' so to speak.

Reply:
Right, there's obviously a tradeoff here, I'll leave it up to you which direction we go (restricting the max during routing more, or making the privacy offset more restricted), but I don't think failing to find a route just because of the privacy offset is an acceptable solution.

Reply:
https://github.com/lightningdevkit/rust-lightning/pull/1286/commits/4e9cb21af935e45bee51d4dee8790459a995e38e is a move in this direction: limit the overall applied offset, but don't fail the pathfinding. 
One question that's still open for me, is, whether we should make sure that the offset is a multiple of `40` or `144` when we do indeed limit it, or whether this degree of obfuscation is not worth the effort.